telemetry/telemetry.go

```diff
@@ -0,0 +1,406 @@
+			resp.Body.Close()
+			continue
+		}
+
+		// read the response body
+		err = json.NewDecoder(resp.Body).Decode(&reply)
+		resp.Body.Close() // close response body as soon as we're done with it
+		if err != nil {
+			continue
+		}
+
+		// update the list of enabled/disabled keys, if any
+		for _, key := range reply.EnableKeys {
+			disabledMetricsMu.Lock()
+			// only re-enable this metric if it is temporarily disabled
+			if temp, ok := disabledMetrics[key]; ok && temp {
+				delete(disabledMetrics, key)
+			}
+			disabledMetricsMu.Unlock()
+		}
+		for _, key := range reply.DisableKeys {
+			disabledMetricsMu.Lock()
+			disabledMetrics[key] = true // all remotely-disabled keys are "temporarily" disabled
+			disabledMetricsMu.Unlock()
+		}
+
+		// make sure we didn't send the update too soon; if so,
+		// just wait and try again -- this is a special case of
+		// error that we handle differently, as you can see
+		if resp.StatusCode == http.StatusTooManyRequests {
+			if reply.NextUpdate <= 0 {
+				raStr := resp.Header.Get("Retry-After")
+				if ra, err := strconv.Atoi(raStr); err == nil {
+					reply.NextUpdate = time.Duration(ra) * time.Second
+				}
+			}
+			if !final {
+				log.Printf("[NOTICE] Sending telemetry: we were too early; waiting %s before trying again", reply.NextUpdate)
+				time.Sleep(reply.NextUpdate)
+				continue
+			}
+		} else if resp.StatusCode >= 400 {
+			err = fmt.Errorf("telemetry server returned status code %d", resp.StatusCode)
+			continue
+		}
+
+		break
+	}
+	if err == nil && !final {
+		// (remember, if there was an error, we return it
+		// below, so it WILL get logged if it's supposed to)
+		log.Println("[INFO] Sending telemetry: success")
+	}
+
+	// even if there was an error after all retries, we should
+	// schedule the next update using our default update
+	// interval because the server might be healthy later
+
+	// ensure we won't slam the telemetry server
+	if reply.NextUpdate < 1*time.Second {
+		reply.NextUpdate = defaultUpdateInterval
+	}
+
+	// schedule the next update (if this wasn't the last one and
+	// if the remote server didn't tell us to stop sending)
+	if !final && !reply.Stop {
+		updateTimerMu.Lock()
+		updateTimer = time.AfterFunc(reply.NextUpdate, func() {
+			logEmit(false)
+		})
+		updateTimerMu.Unlock()
+	}
+
+	return err
+}
+
+func stopUpdateTimer() {
+	updateTimerMu.Lock()
+	updateTimer.Stop()
+	updateTimer = nil
+	updateTimerMu.Unlock()
+}
+
+// makePayloadAndResetBuffer prepares a payload
+// by emptying the collection buffer. It returns
+// the bytes of the payload to send to the server.
+// Since the buffer is reset by this, if the
+// resulting byte slice is lost, the payload is
+// gone with it.
+func makePayloadAndResetBuffer() ([]byte, error) {
+	bufCopy := resetBuffer()
+
+	// encode payload in preparation for transmission
+	payload := Payload{
+		InstanceID: instanceUUID.String(),
+		Timestamp:  time.Now().UTC(),
```

Comment:
Don't trust this time. Instead, get it from the TLS handshake with the telemetry endpoint.

Unless, of course, this is for a survey of clock drift(s).

Reply:
* Be prepared for a new UUID every time a throwaway instance is spun up.
 * Keep in mind that some operators will constrain the process `caddy`, not allow any outgoing connections (other than to a few ACME hosts perhaps) or/and writing to disk (does it break the part that generates the UUID? I didn't check).

Reply:
Does the Go standard lib expose the timestamp of the server contained in the TLS handshake?

And, the contingencies you mentioned are acceptable I think. (And if the UUID can't be written to disk, it'll continue gracefully but with a log entry.)

Reply:
Re time: come to think of it, leave as-is and just log with the endpoint's local datetime. A survey of clock drift is a good thing to have given how many short-lived tokens are used.

Reply:
Sounds good, I think I agree!